### PR TITLE
Postpone Composite name creating

### PR DESCRIPTION
**Pull request by nouiz**
Fix some compilation getting killed.

I does this. The error was related to the elemwise fusion. I postpone some work from Composite.__init__ to later when we use it as all the stack trace we got when doing CTRL-C during the execution was there. This should speed up compilation, but it also fixed the crashes.

### CODE REVIEW

File: theano/tensor/elemwise.py [hunk 1]
```diff
@@ -849,6 +849,14 @@ def prepare_node(self, node, storage_map, compute_map):
...
             sig = char * node.nin + '->' + char * node.nout
             node.tag.sig = sig
+        node.tag.fake_node = Apply(
+            self.scalar_op,
+            [get_scalar_type(dtype=input.type.dtype).make_variable()
+             for input in node.inputs],
+            [get_scalar_type(dtype=output.type.dtype).make_variable()
+             for output in node.outputs])
+
+        self.scalar_op.prepare_node(node.tag.fake_node, [], [])
```
Comment (abergeron) [inline theano/tensor/elemwise.py]: At least pass None for the fake storage_map and compute_map

Comment (nouiz) [inline theano/tensor/elemwise.py]: done

File: theano/scalar/basic.py [hunk 2]
```diff
@@ -3538,21 +3538,38 @@ def init_py_impls(self):
...
+        # the memo is to reuse functions during this call and traverse
+        # the graph only once. We will still have the problem during
+        # execution for the next comment. But this speed up linker
+        # phase where this method is called. Not so big problem as
+        # this code isn,t when performance is important.
+        memo = {}
+
         def compose_impl(r):
             # this is not optimal at all eg in add(*1 -> mul(x, y), *1)
             # it will calculate *1 twice
```
Comment (abergeron) [inline theano/scalar/basic.py]: This isn't true anymore

Comment (nouiz) [inline theano/scalar/basic.py]: This is True for execution. I'll try to clarify.

**Comment by nouiz** [comment]
I updated the PR with more optimization and the fix from code review and CI failure.

File: theano/tensor/elemwise.py [hunk 3]
```diff
@@ -849,6 +849,14 @@ def prepare_node(self, node, storage_map, compute_map):
             char = numpy.sctype2char(out_dtype)
             sig = char * node.nin + '->' + char * node.nout
             node.tag.sig = sig
+        node.tag.fake_node = Apply(
+            self.scalar_op,
+            [get_scalar_type(dtype=input.type.dtype).make_variable()
+             for input in node.inputs],
+            [get_scalar_type(dtype=output.type.dtype).make_variable()
+             for output in node.outputs])
```
Comment (abergeron) [inline theano/tensor/elemwise.py]: This should probably use make_node()

Comment (nouiz) [inline theano/tensor/elemwise.py]: This is a fake node. This is faster then calling make_node as it don't do many checks. I don't want to run the risk of slowing that down compared to the old code. I took it from line 1112 in the old version.

**Pull request merged**